api/stocks.go

```diff
@@ -113,11 +210,199 @@ func (s *Server) fetchStockData(ctx context.Context, symbol string) (*TimeSeries
+		resultCh <- fetchResult{overview: overview, err: err}
+	}()
+
+	go func() {
+		defer wg.Done()
+		balanceSheet, err := h.fetchBalanceSheet(ctx, symbol)
+		resultCh <- fetchResult{balanceSheet: balanceSheet, err: err}
+	}()
+
+	go func() {
+		wg.Wait()
+		close(resultCh)
+	}()
+
+	var overview *OverviewMetadata
+	var balanceSheet *BalanceSheetMetadata
+	var fetchErrors []error
+
+	for res := range resultCh {
+		if res.err != nil {
+			fetchErrors = append(fetchErrors, res.err)
+		}
+		if res.overview != nil {
+			overview = res.overview
+		}
+
+		if res.balanceSheet != nil {
+			balanceSheet = res.balanceSheet
+		}
+	}
+
+	if len(fetchErrors) == numCh {
+		return nil, nil, fmt.Errorf("failed to fetch stock data: %w & %w", fetchErrors[0], fetchErrors[1])
+	}
+
+	return overview, balanceSheet, nil
+}
+
+func processBalanceSheetMetrics(
+	ctx context.Context,
+	balanceSheet *BalanceSheetMetadata,
+	metricMap map[string]model.Metric,
+	save func(model.Metric, float64),
+	symbol string,
+) {
+	if metric, ok := metricMap["Debt/Equity Ratio"]; ok {
+		value, err := calculateDebtEquityRatio(ctx, balanceSheet)
+		if err != nil {
+			log.Printf("failed to calculate Debt/Equity Ratio for stock %s: %v", symbol, err)
+			return
+		}
+		save(metric, value)
+	}
+}
+
+func processOverviewMetrics(
+	_ context.Context,
+	overview *OverviewMetadata,
+	metricMap map[string]model.Metric,
+	save func(model.Metric, float64),
+) {
+	metricExtractors := map[string]func(*OverviewMetadata) string{
+		"P/E Ratio":      func(o *OverviewMetadata) string { return o.PERatio },
+		"EPS":            func(o *OverviewMetadata) string { return o.EPS },
+		"Dividend Yield": func(o *OverviewMetadata) string { return o.DividendYield },
+		"Market Cap":     func(o *OverviewMetadata) string { return o.MarketCapitalization },
+		"Revenue Growth": func(o *OverviewMetadata) string { return o.QuarterlyRevenueGrowthYOY },
+	}
+
+	for name, metricModel := range metricMap {
+		if extractor, exists := metricExtractors[name]; exists {
+			value, err := strconv.ParseFloat(extractor(overview), 64)
+			if err != nil {
+				log.Printf("failed to convert metric %s to float: %v", name, err)
+			}
+			save(metricModel, value)
+		}
+	}
+}
+
+func (h *StockHandler) metricAdder(
+	ctx context.Context,
+	stock *model.Stock,
+	updatedMetrics *[]model.StockMetric,
+) func(metric model.Metric, value float64) {
+	return func(metric model.Metric, value float64) {
+		stockMetric := &model.StockMetric{
+			StockID:    stock.ID,
+			MetricID:   metric.ID,
+			Value:      value,
+			RecordedAt: time.Now(),
+		}
+		savedMetric, err := h.store.CreateStockMetric(ctx, stockMetric)
+		if err != nil {
+			log.Printf("failed to save metric %s for stock %s: %v", metric.Name, stock.Symbol, err)
+			return
+		}
+		*updatedMetrics = append(*updatedMetrics, *savedMetric)
+	}
+}
+
+func buildMetricMap(metrics []model.Metric) map[string]model.Metric {
+	metricMap := make(map[string]model.Metric)
```

Comment:
```suggestion
	metricMap := make(map[string]model.Metric, len(metrics))
```

Pre-allocate the map